minecraft/networking/packets.py

```diff
@@ -727,6 +727,399 @@ def __str__(self):
+        self.radius = Float.read(file_object)
+        records_count = VarInt.read(file_object)
+        self.records = []
+        for i in range(records_count):
+            rec_x = Byte.read(file_object)
+            rec_y = Byte.read(file_object)
+            rec_z = Byte.read(file_object)
+            record = ClientExplosion.Record(rec_x, rec_y, rec_z)
+            self.records.append(record)
+        self.player_motion_x = Float.read(file_object)
+        self.player_motion_y = Float.read(file_object)
+        self.player_motion_z = Float.read(file_object)
+
+    def write(self, socket, compression_threshold=None):
+        raise NotImplementedError
+
+
+class ClientSpawnObject(Packet):
+    @staticmethod
+    def get_id(context):
+        return 0x00 if context.protocol_version >= 67 else \
+               0x0E
+
+    packet_name = 'spawn object'
+
+    class Type(object):
+        def read(self, file_object):
+            self._read(file_object)
+
+        def _read(self, file_object):
+            raise NotImplementedError(
+                'This abstract method must be overridden in a subclass.')
+
+        @classmethod
+        def type_from_id(cls, type_id):
+            subcls = {
+                1: ClientSpawnObject.Type_Boat,
+                2: ClientSpawnObject.Type_Item_Stack,
+                3: ClientSpawnObject.Type_Area_Effect_Cloud,
+                10: ClientSpawnObject.Type_Minecart,
+                50: ClientSpawnObject.Type_Activated_TNT,
+                51: ClientSpawnObject.Type_EnderCrystal,
+                60: ClientSpawnObject.Type_Arrow,
+                61: ClientSpawnObject.Type_Snowball,
+                62: ClientSpawnObject.Type_Egg,
+                63: ClientSpawnObject.Type_FireBall,
+                64: ClientSpawnObject.Type_FireCharge,
+                65: ClientSpawnObject.Type_Enderpearl,
+                66: ClientSpawnObject.Type_Wither_Skull,
+                67: ClientSpawnObject.Type_Shulker_Bullet,
+                68: ClientSpawnObject.Type_Llama_spit,
+                70: ClientSpawnObject.Type_Falling_Objects,
+                71: ClientSpawnObject.Type_Item_frames,
+                72: ClientSpawnObject.Type_Eye_of_Ender,
+                73: ClientSpawnObject.Type_Potion,
+                75: ClientSpawnObject.Type_Exp_Bottle,
+                76: ClientSpawnObject.Type_Firework_Rocket,
+                77: ClientSpawnObject.Type_Leash_Knot,
+                78: ClientSpawnObject.Type_ArmorStand,
+                79: ClientSpawnObject.Type_Evocation_Fangs,
+                90: ClientSpawnObject.Type_Fishing_Hook,
+                91: ClientSpawnObject.Type_Spectral_Arrow,
+                93: ClientSpawnObject.Type_Dragon_Fireball
+            }.get(type_id)
+            if subcls is None:
+                raise ValueError("Unknown type ID: %s."
+                                 % type_id)
+            return subcls
+
+    class Type_Boat(Type):
```

Comment:
I think you followed the example from the `PlayerListItemPacket` for this, the reason it has this subclassing is because the packet varies depending on the type of Action. In this case, the type doesn't vary so you can forego this and simply introduce an enum like thing that looks like:

```python
class EntityType:
    BOAT = 1
    ITEM_STACK = 2
    # ...

    by_id = {id: entity for (entity, id) in EntityType.__dict__.items() if entity.isupper()}

    @staticmethod
    def get_entitity_type_by_id(id):
        return self.by_id[id]
```